src/System.Collections/src/System/Collections/Generic/SortedSet.cs

```diff
@@ -1132,13 +1132,14 @@ internal virtual void IntersectWithEnumerable(IEnumerable<T> other)
                 if (Contains(item))
                 {
                     toSave.Add(item);
+                    DoRemove(item);
                 }
             }
 
-            if (toSave.Count < Count)
+            Clear();
+            foreach (T item in toSave)
             {
-                Clear();
-                AddAllElements(toSave);
+                AddIfNotPresent(item);
```

Comment:
Changing AddAllElements would break the logic as it's used from `UnionWith` which needs to check for `Contains` before adding. I could inline the helper if you want as it's used only from there now